File: ios/BetterLise/Features/Grades/GradeSorting.swift
import Foundation

enum GradeSorting {
    /// Unread grades first, then most recent Lise date first.
    static func sorted(_ grades: [Grade]) -> [Grade] {
        grades.sorted { lhs, rhs in
            if lhs.isUnread != rhs.isUnread { return lhs.isUnread }
            let lhsDate = DateParsing.liseDay(lhs.date) ?? .distantPast
            let rhsDate = DateParsing.liseDay(rhs.date) ?? .distantPast
            return lhsDate > rhsDate
        }
    }

    enum Tier: Equatable {
        case failing, passing, good
    }

    static func tier(for note: Double) -> Tier {
        switch note {
        case ..<10: .failing
        case ..<12: .passing
        default: .good
        }
    }

    /// Index of the distribution bin (2-point bins, 20 counted in the last one).
    static func binIndex(for note: Double, binCount: Int = 10) -> Int {
        min(max(Int(note / 2), 0), binCount - 1)
    }
}

File: android/app/src/main/java/com/betterlise/app/domain/GradeSorting.kt
package com.betterlise.app.domain

import com.betterlise.app.data.api.Grade
import java.time.LocalDate

object GradeSorting {
    enum class Tier { Failing, Passing, Good }

    /** Unread grades first, then most recent Lise (dd/MM/yyyy) date first. */
    fun sorted(grades: List<Grade>): List<Grade> = grades.sortedWith(
        compareByDescending<Grade> { it.isUnread }
            .thenByDescending { parseLiseDate(it.date) ?: LocalDate.MIN },
    )

    fun parseLiseDate(value: String): LocalDate? {
        val parts = value.split("/").mapNotNull { it.toIntOrNull() }
        if (parts.size != 3) return null
        return runCatching { LocalDate.of(parts[2], parts[1], parts[0]) }.getOrNull()
    }

    fun tier(note: Double): Tier = when {
        note < 10 -> Tier.Failing
        note < 12 -> Tier.Passing
        else -> Tier.Good
    }

    /** Index of the distribution bin (2-point bins, 20 counted in the last one). */
    fun binIndex(note: Double, binCount: Int = 10): Int = (note / 2).toInt().coerceIn(0, binCount - 1)
}
